### ISSUE TITLE
chassis.cpp is taking 6-11 minutes to compile

### ISSUE THREAD

**Issue by randomascii** [KhronosGroup/Vulkan-ValidationLayers#2889]
**Describe the Issue**
While working on Chrome I recently noticed that chassis.cpp recently started taking 6-11 minutes to compile on Windows. The exact time depends on the compiler flags and on how busy the machine is. On optimized builds using clang-cl the compile time for just this file is about six minutes when it gets a core to itself, and about eleven minutes when it is sharing a core with other tasks.

**Environment:**
 - OS: Windows
 - GPU: n/a
 - SDK or header version if building from repo: "commit": known_good.json says commit "v1.2.180" for Vulkan-Headers. Beyond that I'm not certain - around one week ago.

This is in a Chromium build on Windows. See also crbug.com/1219705. I assume that other environments are also hitting long compilation times with this file.

**Comment by ncesario-lunarg** [comment]
Thanks for the PR @randomascii, could you please share the compiler flags you're using? 6-11 minutes sounds about how long the entire build should take :).

chassis.cpp is [generated](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/ae3d0c3217316cd0364d4b9069d0e2afac5957b1/scripts/layer_chassis_generator.py#L63) and could be broken up into multiple cpp files, but it would require some refactoring of the generator code.

**Comment by randomascii** [comment]
When I'm doing a full Chrome build the cost of compiling this file usually manages to get overlapped with enough other steps (I use our build farm and get 1,000 way parallelism) to not affect things too much, but if something triggers a rebuild of chassis.obj when not much is going on it can make the incremental rebuild take ~ten times longer than it might otherwise. On a recent build chassis.obj took 683.6 s and the entire build of Chrome (49,972 build steps) took 1148.6 s.

Anyway, here is some relevant output from ClangBuildAnalyzer after doing a build of Chrome with the gn args listed above:

**** Files that took longest to parse (compiler frontend):
 76452 ms: out/win10/obj/third_party/vulkan-deps/vulkan-validation-layers/src/VkLayer_khronos_validation/chassis.obj
 61754 ms: out/win10/obj/third_party/blink/renderer/bindin
...
**** Files that took longest to codegen (compiler backend):
650325 ms: out/win10/obj/third_party/vulkan-deps/vulkan-validation-layers/src/VkLayer_khronos_validation/chassis.obj
 45201 ms: out/win10/obj/third_party/perfetto/src/trace_processor/db/db/column.obj
...
**** Functions that took longest to compile:
147158 ms: public: void __cdecl ValidationObject::InitObjectDispatchVectors(void) (../../third_party/vulkan-deps/vulkan-validation-layers/src/layers/generated/chassis.cpp)
  5640 ms: ValidateChunkAMD64 (../../native_client/src/trusted/validator_ragel/gen/validator_x86_64.c)

chassis.cpp is the most expensive source file in three different categories. The code-gen time is particularly impressive. The Chromium bug (crbug.com/1219705) has an attached chassis.json file that can be loaded into chrome://tracing to see exactly where time is going. That file was generated by compiling with -ftime-trace

Here are the compile options that I have been using lately. This is for an x64 release build. This compile took 444.5 s (faster due to not sharing a core with other compiles):

..\..\third_party\llvm-build\Release+Asserts\bin\clang-cl.exe /nologo /winsysroot../../third_party/depot_tools/win_toolchain/vs_files/20d5f2553f -DNOMINMAX -DDCHECK_ALWAYS_ON=1 -DUSE_AURA=1 "-DCR_CLANG_REVISION=\"llvmorg-13-init-12881-g4017d033-1\"" -D_LIBCPP_ABI_UNSTABLE -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_NODISCARD -D_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS -D_LIBCPP_DEBUG=0 -D_LIBCPP_NO_AUTO_LINK -D__STD_C -D_CRT_RAND_S -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D_ATL_NO_OPENGL -D_WINDOWS -DCERT_CHAIN_PARA_HAS_EXTRA_FIELDS -DPSAPI_VERSION=2 -DWIN32 -D_SECURE_ATL -DWINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP -DWIN32_LEAN_AND_MEAN -DNTDDI_VERSION=NTDDI_WIN10_VB -D_WIN32_WINNT=0x0A00 -DWINVER=0x0A00 -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DVULKAN_NON_CMAKE_BUILD "-DAPI_NAME=\"Vulkan\"" -DVK_ENABLE_BETA_EXTENSIONS -DVK_USE_PLATFORM_WIN32_KHR -I../.. -Igen -I../../third_party/vulkan-deps/vulkan-validation-layers/src/layers -I../../third_party/vulkan-deps/vulkan-validation-layers/src/layers/generated -I../../third_party/vulkan-deps/vulkan-headers/src/include -I../../third_party/vulkan-deps/glslang/src -I../../third_party/vulkan-deps/spirv-tools/src/include -I../../third_party/vulkan-deps/spirv-headers/src/include -fno-delete-null-pointer-checks -fno-ident -fcolor-diagnostics -fmerge-all-constants -fcrash-diagnostics-dir=../../tools/clang/crashreports -mllvm -instcombine-lower-dbg-declare=0 -fcomplete-member-pointers /Gy /FS /bigobj /utf-8 /Zc:twoPhase /Zc:sizedDealloc- /D__WRL_ENABLE_FUNCTION_STATICS__ -fmsc-version=1916 -m64 -msse3 /Brepro -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -Xclang -fdebug-compilation-dir -Xclang . -no-canonical-prefixes /O1 /Ob2 /Oy- /Zc:inline /Gw /Oi /Z7 -gcodeview-ghash -ftrivial-auto-var-init=pattern /guard:cf,nochecks /MT -Xclang -add-plugin -Xclang find-bad-constructs -Xclang -plugin-arg-find-bad-constructs -Xclang checked-ptr-as-trivial-member -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare /W3 /wd4800 /wd4267 /wd4996 -Wno-unused-variable -Wno-misleading-indentation /WX /wd4091 /wd4127 /wd4251 /wd4275 /wd4312 /wd4324 /wd4351 /wd4355 /wd4503 /wd4589 /wd4611 /wd4100 /wd4121 /wd4244 /wd4505 /wd4510 /wd4512 /wd4610 /wd4838 /wd4995 /wd4996 /wd4456 /wd4457 /wd4458 /wd4459 /wd4200 /wd4201 /wd4204 /wd4221 /wd4245 /wd4267 /wd4305 /wd4389 /wd4702 /wd4701 /wd4703 /wd4661 /wd4706 /wd4715 -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-nonportable-include-path -Wno-psabi -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-final-dtor-non-final-class -Wno-builtin-assume-aligned-alignment -Wno-deprecated-copy -Wno-non-c-typedef-for-linkage -Wno-unused-but-set-parameter -Wno-unused-but-set-variable -Wmax-tokens -Wno-unused-function -Wno-trigraphs /TP /GR- -I../../buildtools/third_party/libc++/trunk/include /wd4577 /GR /c ../../third_party/vulkan-deps/vulkan-validation-layers/src/layers/generated/chassis.cpp /Foobj/third_party/vulkan-deps/vulkan-validation-layers/src/VkLayer_khronos_validation/chassis.obj /Fd"obj/third_party/vulkan-deps/vulkan-validation-layers/src/VkLayer_khronos_validation_cc.pdb"

**Comment by jeremyg-lunarg** [comment]
seems like making chassis_dispatch_helper.h into a .cpp would be pretty easy and would at least get the really slow function into its own file that could churn in parallel with others.

**Comment by aeubanks** [comment]
I tried wrapping the BUILD_DISPATCH_VECTOR macro in a lambda and that cuts down the time to build chassis.cpp from 2m35s to 1m25s. Otherwise ValidationObject::InitObjectDispatchVectors() is basically a 40000 line function, which some compilers don't do well with.

Would you be open to a change like that?

For reference, currently the code is something like
```
#define BUILD_DISPATCH_VECTOR(name) \
 ... // a loop which is 40 lines of code

ValidationObject::InitObjectDispatchVectors() {
 BUILD_DISPATCH_VECTOR(foo);
 // 1000 more of these
}
```

**Comment by randomascii** [comment]
A lambda instead of a macro sounds like an improvement even without the build-speed improvement. Does using a lambda also result in a smaller .obj file? Currently chassis.obj is ~33 MB (release 64-bit build on Windows) and a lot of that could be code replication due to the macro which I assume the lambda avoids.

**Comment by aeubanks** [comment]
hmm, the size actually increases with my patch from 14.2MB to 14.5MB. Somehow the typeid checks were previously getting optimized away, but now they're not.
But my newer better patch actually makes the time go from 2m34s to 33s. The previous iteration was still creating 1000 versions of the function, now we just have 1 function.

**Comment by randomascii** [comment]
I noticed that aeubanks' build times are much smaller than mine and I wondered if that might be related to optimization, since I have been testing on a release (optimized) build. I added "#pragma clang optimize off" and the compile time dropped from 420 s to 54 s. So, that's another option I guess, if this code isn't performance critical.

That change made the .obj file grow from 33 MB to 42 MB.

The lambda technique would be better than disabling optimizations.
